### PR TITLE
adapter: Allow canceling linearized reads

### PR DESCRIPTION
**Pull request by jkosh44**
If a read is executed at a timestamps larger than the timestamp oracle's read timestamp, then that read must be linearized. Linearized here means that the read must wait for the timestamp oracle's read timestamp to catch up to the timestamp of the read before returning a result to the client. This situation is very rare and should only happen when the since of some object goes past the oracle's read timestamp, which should only happen due to clock skew.

Previously, while a read was being linearized (i.e. waiting), it was impossible to cancel to read query. This commit fixes that issue by allowing read queries to be canceled while they're being linearized.

The bulk of the work was storing the pending read transactions in a field of the Coordinator instead of passing them around in channels and tasks.

### Motivation
This PR fixes a previously unreported bug.

### Checklist

- [X] This PR has adequate test coverage / QA involvement has been duly considered.
- [X] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [X] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [X] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [X] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - Allow canceling read queries more often.

### CODE REVIEW

**Comment by jkosh44** [comment]
I have no idea how write a test for this. I was only only able to discover it due to this branch: https://github.com/MaterializeInc/materialize/pull/24704, which caused the since of a Materialized view to be far in the future (which may or may not be a bug).

**Comment by jkosh44** [comment]
@ParkMyCar I'm going to rebase over https://github.com/MaterializeInc/materialize/pull/24704 so that I can add a test. In the meantime I'll mark this as draft until the other PR (which is set to auto-merge) merges. I'm not actually changing the contents of the commit in this PR at all if you've already started the review.

File: src/environmentd/tests/sql.rs
Comment (ParkMyCar) [inline src/environmentd/tests/sql.rs:3835]: 😂

**Review by ParkMyCar — APPROVED** [review]
Do we think this was the cause of this envd getting stuck? [Slack](https://materializeinc.slack.com/archives/C06FG1186J2/p1706213431508159?thread_ts=1706213335.134189&cid=C06FG1186J2)

Nice!

**Comment by jkosh44** [comment]
> Do we think this was the cause of this envd getting stuck? [Slack](https://materializeinc.slack.com/archives/C06FG1186J2/p1706213431508159?thread_ts=1706213335.134189&cid=C06FG1186J2) 

Unlikely, this can cause a session to get stuck but not the entire Coordinator. The waiting is still done off of the main thread.

**Comment by jkosh44** [comment]
The test failure is very interesting, it seems to suggest that after cancelling the query, we still send back a response. I'm looking into it.

**Comment by jkosh44** [comment]
> The test failure is very interesting, it seems to suggest that after cancelling the query, we still send back a response. I'm looking into it.

Ok, so what's happening is something like the following:

1. Coord receives peek.
2. Coord executes peek.
3. Coord sends peek results to client.
4. Coord receives commit.
5. Coord linearizes txn (waits for timestamp oracle to catch up to txn timestamp).
6. Txn is cancelled.
7. Coord sends back Canceled response.

The test was originally passing locally because there's a race for when the cancel is received. If it's received before step 3 the test will pass.

I'll have to think about the right approach here. Cancelling should probably also abort the transaction, if it's not already doing that.

**Comment by jkosh44** [comment]
After further inspection, I don't think read-then-write cancelling works correctly either. The read-then-write query will wait forever on a response from it's read channel which it will never get.

**Comment by jkosh44** [comment]
I figured out what was wrong, but this PR is now blocked on this PR: https://github.com/MaterializeInc/materialize/pull/24862

**Comment by jkosh44** [comment]
@ParkMyCar Ok sorry again to drag this out but this is finally read for review again. All the commits but the first one are new, those are

- https://github.com/MaterializeInc/materialize/pull/24826/commits/d97ea6882d0042a3060f66424306848947b83865
- https://github.com/MaterializeInc/materialize/pull/24826/commits/c229c30b4e688735f2f875d93e044d676c015b40
- https://github.com/MaterializeInc/materialize/pull/24826/commits/726be9bad93e141d234e7c8420dec20b4c738758
- https://github.com/MaterializeInc/materialize/pull/24826/commits/7ca32f46bd18c44287a203aca7bf2b40c5c12e8f
- https://github.com/MaterializeInc/materialize/pull/24826/commits/ec2f34b3b11904905f21d0c7565bffdf2e8081fe
- https://github.com/MaterializeInc/materialize/pull/24826/commits/2a503df988a65c9b45370ec3c15e884256e3faf2

I'm going to leave some comments inline too that explain some of the more impactful changes.

File: src/adapter/src/coord/command_handler.rs
Comment (jkosh44) [inline src/adapter/src/coord/command_handler.rs:922]: This IMO is the biggest difference. We are able to respond both with an `Ok(ExecuteResponse::Canceled)` and an `Err(AdapterError::Canceled)`. They have different semantics with respect to what the client ultimately receives that I don't think I full understand, however I think sending an `Err` here is the right approach. I'm going to post a comment on a specific test explaining why.

File: src/environmentd/tests/server.rs
Comment (jkosh44) [inline src/environmentd/tests/server.rs:2832]: This is the only test that started failing as a result of the change here: https://github.com/MaterializeInc/materialize/pull/24826/files#r1481988333 (switching from sending `Ok(ExecuteResponse::Canceled)` to `Err(AdapterError::Canceled)`).

As you can see the big difference is that the client used to receive an `Ok` response and now they receive an `Err` response. IMO `Err` is the correct semantics here. If you read this test you'll see that previously it was expecting this `INSERT` to both fail and send an `Ok` response, which doesn't make a lot of sense. Now it fails and sends an `Err` response which makes much more sense to me.

File: src/environmentd/tests/sql.rs
Comment (jkosh44) [inline src/environmentd/tests/sql.rs:3878]: This test was combined with `test_cancel_linearize_reads`. Previously, this test would start a query that never completes. `ClientTransmitter` has a panic in it's drop implementation if a response was never sent to the client. Previously the `ClientTransmitter` lived in an async task that would eat the panic and the test would still pass. Now the `ClientTransmitter` transmitter is stored as a field in the `Coordinator`, so dropping the `Coordinator` also drops the `ClientTransmitter`, which panics and fails the test.

**Pull request merged**
